### PR TITLE
Expose reverse tunnels to users

### LINKED ISSUE #437: Document NAT behind-firewall / reverse tunnel operation
This issue is about @klizhentas' [comment](https://github.com/gravitational/teleport/issues/309#issuecomment-204829087) in #309:
>  Teleport has first class support for reverse tunnels just for this special case. We will elaborate in the docs to address this particular problem.

#309 was really two issues:

1. Originally it was about `--advertise-ip` which wasn't well documented
2. I piggy-backed that issue to also ask about reverse tunnels / operating behind firewalls/NAT.

The first sub-issue about `--advertise-ip` has been resolved in #309 which therefore has been closed, but the second sub-issue about reverse tunnels has not. This issue covers this second sub-issue

### LINKED ISSUE #443: Working through the proxy and firewall
Hi!

Wonderful idea, guys! 
But i don't understand if teleport can help me or not. i have such a scheme 

1. server with public ip. I can control it by myself in any case. lets name it remote
2. several servers behind the firewall and proxy. from them I have only access to internet to port http/https. lets call them cluster.
3. i have vpn from remote to cluster. 
4. i can make connections from remote to cluster to port 22/443. others are blocked. 
5. on remote side on port 443 i already have web site behind the nginx. so i can proxy requests to teleport

So my question is, can teleport help me get access from internet to cluster?

### PR DESCRIPTION
**Pull request by kontsevoy**
This PR adds:
- New configuration setting: "trusted_clusters" in teleport.yaml
- Documentation for it
- "Nomenclature" section for online docs
- `clusters` command for `tsh`
- `--cluster` flag for `tsh`

This PR fixes & changes:
- `tctl auth ls` command now shows local vs "trusted" clusters.
- `tctl auth export` uses different format for exporting CA keys, and it only exports local keys.
- Better user-facing error messages and more user-friendly logging.
- A couple of configuration bugs.

Fixes #437 
Fixes #443 
### Potentially Breaking Changes
1. `tctl auth export` now uses "known_hosts" format for both user and host CAs. Does this break exporting keys and using them with OpenSSH?
2. "auth_servers" configuration value now uses "host:port" format instead of "tcp://host:port". 
### Undocumented Changes

It is now possible to specify `*` (wildcard) as "allowed logins" for a user CA. This is how you can say "trust every key signed by this CA". I have not documented it yet.

### CODE REVIEW

File: docs/admin-guide.md [hunk 1]
```diff
@@ -30,9 +30,22 @@ this way:
...
-## Running
+## Nomenclature
 
-Teleport supports only a handful of commands
+Before diving into configuring and running Teleport, it helps to take a look at the [Teleport Architecture](/architecture) 
+and go over the key concepts this document will be referring to:
+
+|Concept   | Description
+|----------|------------
+|Node      | Synonym to "server" or "computer", something one can "SSH to". A node must be running `teleport` daemon running with "node" role/service turned on.
```
Comment (klizhentas) [inline docs/admin-guide.md]: or container

File: docs/admin-guide.md [hunk 2]
```diff
@@ -451,6 +464,124 @@ Node Name     Node ID          Address         Labels
...
+
+Lets assume we need to place some servers behind a firewall, and we only want Teleport 
+user "john" to have access to them. Assume we already have our primary Teleport cluster 
+and our users set up. Say, this cluster is called "main". 
+
+Now, to add behind-the-firewall machines and restrict access only to "john", we will have 
+to do the following:
+
+1. Create a new cluster for behind-the-firewall machines. Lets call it "cluster-b".
+2. Add "cluster-b" to the list of `trusted clusters` of "main-cluster".
```
Comment (klizhentas) [inline docs/admin-guide.md]: of `main` not `main-cluster`  as config's cluster name is `main`

File: docs/admin-guide.md [hunk 3]
```diff
@@ -451,6 +464,124 @@ Node Name     Node ID          Address         Labels
...
+Now, modify the YAML configuration of both clusters to connect them.
+
+On "main" (running on 62.28.10.1)
+
+```yaml
+auth_service:
+  enabled: yes
+  cluster_name: main
+  trusted_clusters:
+      - keyfile: /path/to/b-cluster.ca
```
Comment (klizhentas) [inline docs/admin-guide.md]: `keyfile` vs `key_file`? I wonder if we have configs somewhere with `_file`

**Comment by klizhentas** [comment]
I'll dig in more detail tomorrow

**Comment by klizhentas** [comment]
tests fail:

```
--- PASS: TestClientAPI (0.69s)
PASS
coverage: 16.6% of statements
ok  	github.com/gravitational/teleport/lib/client	1.771s	coverage: 16.6% of statements
=== RUN   TestConfig

----------------------------------------------------------------------
FAIL: config_test.go:216: ConfigTestSuite.TestTrustedClusters

config_test.go:228:
    c.Assert(err, check.IsNil)
... value *trace.TraceErr = &trace.TraceErr{Err:(*errors.errorString)(0xc82025c880), Traces:trace.Traces{trace.Trace{Path:"github.com/gravitational/teleport/lib/config/_test/_obj_test/configuration.go", Func:"github.com/gravitational/teleport/lib/config.readTrustedClusters", Line:494}}, Message:"", DebugMessage:""} ("reading trusted cluster keys: open ../../fixtures/trusted_clusters/cluster-a: no such file or directory")

OOPS: 5 passed, 1 FAILED
```

File: docs/admin-guide.md (hunk 3 above)
Comment (kontsevoy) [inline docs/admin-guide.md]: good catch. changed to key_file.

File: lib/auth/init.go [hunk 4]
```diff
@@ -174,15 +173,15 @@ func Init(cfg InitConfig) (*AuthServer, *Identity, error) {
...
 		}
 	}
-	if firstStart {
-		if len(cfg.ReverseTunnels) != 0 {
-			log.Infof("FIRST START: Initializing reverse tunnels")
-			for _, tunnel := range cfg.ReverseTunnels {
-				if err := asrv.UpsertReverseTunnel(tunnel, 0); err != nil {
-					return nil, nil, trace.Wrap(err)
-				}
+	if len(cfg.ReverseTunnels) != 0 {
```
Comment (klizhentas) [inline lib/auth/init.go]: so if there's no `firstStart` logic here it means there's no way for me to preconfigure tunnels and then later remove them.

File: lib/config/configuration.go [hunk 5]
```diff
@@ -122,7 +126,11 @@ func ApplyFileConfig(fc *FileConfig, cfg *service.Config) error {
 	if len(fc.AuthServers) > 0 {
 		cfg.AuthServers = make([]utils.NetAddr, 0, len(fc.AuthServers))
 		for _, as := range fc.AuthServers {
-			addr, err := utils.ParseAddr(as)
```
Comment (klizhentas) [inline lib/config/configuration.go]: `utils.ParseAddr` supports both `tcp://` and simple `host:port` format, no need to introduce a breaking change here

File: lib/config/configuration.go [hunk 6]
```diff
@@ -321,6 +329,111 @@ func ApplyFileConfig(fc *FileConfig, cfg *service.Config) error {
...
+// readTrustedClusters parses the content of "trusted_clusters" YAML structure
+// and modifies Teleport 'conf' by adding "authorities" and "reverse tunnels"
+// to it
+func readTrustedClusters(clusters []TrustedCluster, conf *service.Config) error {
+	if len(clusters) == 0 {
+		return nil
+	}
+	// parseCAKey() gets called for every line in a "CA key file" which is
+	// the same as 'known_hosts' format for openssh
+	parseCAKey := func(bytes []byte) (*services.CertAuthority, error) {
```
Comment (klizhentas) [inline lib/config/configuration.go]: can you split this function into several? it's really hard to read as it spans several screens

File: lib/config/configuration.go [hunk 7]
```diff
@@ -321,6 +329,111 @@ func ApplyFileConfig(fc *FileConfig, cfg *service.Config) error {
...
+		marker, options, pubKey, comment, _, err := ssh.ParseKnownHosts(bytes)
+		if marker != "cert-authority" {
+			return nil, trace.Errorf("invalid file format. expected '@cert-authority` marker")
```
Comment (klizhentas) [inline lib/config/configuration.go]: can you use `BadParameter` here and below?

File: lib/config/configuration.go [hunk 8]
```diff
@@ -321,6 +329,111 @@ func ApplyFileConfig(fc *FileConfig, cfg *service.Config) error {
...
+		}
+		if err != nil {
+			return nil, trace.Errorf("invalid public key")
+		}
+		teleportOpts, err := url.ParseQuery(comment)
+		if err != nil {
+			return nil, trace.Errorf("invalid key comment: '%s'", comment)
+		}
+		authType := services.CertAuthType(teleportOpts.Get("type"))
```
Comment (klizhentas) [inline lib/config/configuration.go]: you can get cert type from the certificate spec, no need to put this in the comments.

File: lib/config/configuration.go [hunk 9]
```diff
@@ -321,6 +329,111 @@ func ApplyFileConfig(fc *FileConfig, cfg *service.Config) error {
...
+			if login != "" {
+				allowedLogins = append(allowedLogins, login)
+			}
+		}
+		// open the key file for this cluster:
+		log.Debugf("reading trusted cluster key file %s", tc.KeyFile)
+		if tc.KeyFile == "" {
+			return trace.Errorf("key_file is missing for a trusted cluster")
+		}
+		f, err := os.Open(tc.KeyFile)
```
Comment (klizhentas) [inline lib/config/configuration.go]: we have pretty sweet `util.ReadFile` can you use it instead of `os.Open`

File: lib/reversetunnel/agent.go [hunk 10]
```diff
@@ -56,13 +59,18 @@ type Agent struct {
...
 func NewAgent(
 	addr utils.NetAddr,
-	domainName string,
+	remoteDomainName string,
+	clientName string,
 	signers []ssh.Signer,
 	clt *auth.TunClient) (*Agent, error) {
 
-	log.Debugf("reversetunnel.NewAgent(%v)", domainName)
+	log.Debugf("reversetunnel.NewAgent %s -> %s", clientName, remoteDomainName)
```
Comment (klizhentas) [inline lib/reversetunnel/agent.go]: I recommend using `log.WithFields(log.Fields{"client-name": clientName, "remoteCluster": remoteDomainName})`

File: lib/reversetunnel/agent.go [hunk 11]
```diff
@@ -92,6 +101,10 @@ func (a *Agent) Close() error {
 // Start starts agent that attempts to connect to remote server part
 func (a *Agent) Start() error {
 	conn, err := a.connect()
+	if err != nil {
+		log.Errorf("Failed to create remote tunnel for %v on %s(%s): %v",
```
Comment (klizhentas) [inline lib/reversetunnel/agent.go]: and no return?

File: lib/services/ca.go [hunk 12]
```diff
@@ -86,7 +86,7 @@ func (c *CertAuthID) Check() error {
 		return trace.Wrap(err)
 	}
 	if !cstrings.IsValidDomainName(c.DomainName) {
-		return trace.BadParameter("'%v' is a bad domain name", c.DomainName)
+		return trace.BadParameter("Identity validation error: '%v' is not a valid FQDN", c.DomainName)
```
Comment (klizhentas) [inline lib/services/ca.go]: error sentences should be lowercase, they will be most likely included in logs:

https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: lib/services/ca.go [hunk 13]
```diff
@@ -130,7 +130,7 @@ func (ca *CertAuthority) Checkers() ([]ssh.PublicKey, error) {
 	for _, keyBytes := range ca.CheckingKeys {
 		key, _, _, _, err := ssh.ParseAuthorizedKey(keyBytes)
 		if err != nil {
-			return nil, trace.Wrap(err)
+			return nil, trace.Errorf("invalid authority public key (len=%d): %v", len(keyBytes), err)
```
Comment (klizhentas) [inline lib/services/ca.go]: can you revendor `trace` and start using `trace.Wrap(err, "user friendly error here")`

File: lib/config/configuration.go (hunk 6 above)
Comment (kontsevoy) [inline lib/config/configuration.go]: will do

File: lib/config/configuration.go (hunk 9 above)
Comment (kontsevoy) [inline lib/config/configuration.go]: TODO: check this out

**Comment by klizhentas** [comment]
lgtm

File: lib/config/configuration.go (hunk 7 above)
Comment (kontsevoy) [inline lib/config/configuration.go]: done

File: lib/reversetunnel/agent.go (hunk 11 above)
Comment (kontsevoy) [inline lib/reversetunnel/agent.go]: yep, the following line creates a goroutine which keeps retrying.

**Pull request merged**